File: app/src/main/java/com/ericho/coupleshare/mvp/data/local/LocationsLocalDataSource.java
package com.ericho.coupleshare.mvp.data.local;

import android.content.Context;
import android.support.annotation.NonNull;

import com.ericho.coupleshare.mvp.Location;
import com.ericho.coupleshare.mvp.data.LocationDataSource;
import com.ericho.coupleshare.util.NullUtil;

import org.xutils.DbManager;
import org.xutils.x;

import java.io.IOException;
import java.util.List;

import timber.log.Timber;

/**
 * Created by steve_000 on 10/6/2017.
 * for project CoupleShare
 * package name com.ericho.coupleshare.mvp
 */

public class LocationsLocalDataSource implements LocationDataSource {

    private static LocationsLocalDataSource INSTANCE ;

    private DbManager.DaoConfig daoConfig;

    private DbManager dbManager;

    private LocationsLocalDataSource() {
        daoConfig = DbConfig.getDefaultConfig();
        dbManager = getDbManager();
    }

    public static LocationsLocalDataSource getInstance(){
        if(INSTANCE == null){
            INSTANCE = new LocationsLocalDataSource();
        }

        return INSTANCE;
    }

    public void destroyInstance(){
        if(INSTANCE == null){
            try {
                INSTANCE.dbManager.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        INSTANCE = null;
    }

    private DbManager getDbManager(){
        return x.getDb(daoConfig);
    }


    @Override
    public void getLocations(@NonNull LoadLocationsCallback callback) {
        try {
            List<Location> locations = dbManager.findAll(Location.class);
            if(NullUtil.isNullOrEmpty(locations)) throw new IOException("location no records");

            callback.onLocationsLoaded(locations);
        }catch (IOException e){
            callback.onDataNotAvailable(e);
        }
    }

    @Override
    public void getLocation(@NonNull Integer locationId, @NonNull GetLocationCallback callback) {
        try {
            Location location = dbManager.findById(Location.class,locationId);
            if(NullUtil.isNull(location)) throw new IOException("location not exists");

            callback.onLocationLoaded(location);
        }catch (IOException e){
            callback.onDataNotAvailable(e);
        }
    }

    @Override
    public void saveLocation(@NonNull Location location,SaveLocationCallback callback) {
        try {
            dbManager.save(location);
            callback.onLocationSave();
        }catch (IOException e){
            callback.onLocationSaveFailure(e);
        }
    }

    @Override
    public void deleteAllLocations() {
        try {
            dbManager.delete(Location.class);
        }catch (IOException e){
            Timber.w(e);
        }
    }

    @Override
    public void deleteLocation(@NonNull Integer locationId) {
        try {
            dbManager.deleteById(Location.class,locationId);
        }catch (IOException e){
            Timber.w(e);
        }
    }
}

File: app/src/main/java/com/ericho/coupleshare/mvp/data/local/LocationsLocalDataSource.kt
package com.ericho.coupleshare.mvp.data.local

import com.ericho.coupleshare.mvp.Location
import com.ericho.coupleshare.mvp.data.LocationDataSource
import com.ericho.coupleshare.util.NullUtil
import org.xutils.DbManager
import org.xutils.x
import timber.log.Timber
import java.io.IOException

/**
 * Created by steve_000 on 8/7/2017.
 * for project CoupleShare
 * package name com.ericho.coupleshare.mvp.data.local
 */
class LocationsLocalDataSource: LocationDataSource {


    private val daoConfig: DbManager.DaoConfig

    private val dbManager: DbManager

    init {
        daoConfig = DbConfig.getDefaultConfig()
        dbManager = x.getDb(daoConfig)
    }


    fun destroyInstance() {
        if (INSTANCE != null) {
            try {
                INSTANCE?.dbManager!!.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        INSTANCE = null
    }

    override fun getLocations(callback: LocationDataSource.LoadLocationsCallback) {
        try {
            val locations = dbManager.findAll(Location::class.java)
            if (NullUtil.isNullOrEmpty(locations)) throw IOException("location no records")

            callback.onLocationsLoaded(locations)
        } catch (e: IOException) {
            callback.onDataNotAvailable(e)
        }

    }


    override fun saveLocation(location: Location, callback: LocationDataSource.SaveLocationCallback) {
        try {
            dbManager.save(location)
            callback.onLocationSave()
        } catch (e: IOException) {
            callback.onLocationSaveFailure(e)
        }

    }

    override fun deleteAllLocations() {
        try {
            dbManager.delete(Location::class.java)
        } catch (e: IOException) {
            Timber.w(e)
        }

    }


    override fun getLocation(locationId: Int, callback: LocationDataSource.GetLocationCallback) {
        try {
            val location = dbManager.findById(Location::class.java, locationId)
            if (NullUtil.isNull(location)) throw IOException("location not exists")

            callback.onLocationLoaded(location)
        } catch (e: IOException) {
            callback.onDataNotAvailable(e)
        }

    }

    override fun deleteLocation(locationId: Int) {
        try {
            dbManager.deleteById(Location::class.java, locationId)
        } catch (e: IOException) {
            Timber.w(e)
        }

    }


    companion object {
        @JvmField
        var INSTANCE : LocationsLocalDataSource? = null
        @JvmStatic
        fun getInstance() :LocationsLocalDataSource {
            if (INSTANCE == null) {
                INSTANCE = LocationsLocalDataSource()
            }

            return INSTANCE as LocationsLocalDataSource
        }
    }
}